extensions/amp-viewer-integration/0.1/highlight-handler.js

```diff
@@ -150,9 +157,41 @@ export class HighlightHandler {
     /** @private {?Array<!Element>} */
     this.highlightedNodes_ = null;
 
-    whenDocumentReady(ampdoc.win.document).then(() => {
-      this.initHighlight_(highlightInfo);
-    });
+    if (
+      'fragmentDirective' in document &&
+      isExperimentOn(ampdoc.win, 'use-text-fragments-for-highlights')
+    ) {
+      ampdoc
+        .whenFirstVisible()
+        .then(() => this.highlightUsingTextFragments_(highlightInfo));
+    } else {
+      whenDocumentReady(ampdoc.win.document).then(() => {
+        this.initHighlight_(highlightInfo);
+      });
+    }
+  }
+
+  /**
+   * @param {!HighlightInfoDef} highlightInfo
+   * @private
+   */
+  highlightUsingTextFragments_(highlightInfo) {
+    const {sentences} = highlightInfo;
+    if (!sentences?.length) {
+      return;
+    }
+    const fragment = sentences
+      .map((text) => 'text=' + encodeURIComponent(text))
+      .join('&');
+    this.updateUrlWithTextFragment_(fragment);
+  }
+
+  /**
+   * @param {string} fragment
+   * @private
+   */
+  updateUrlWithTextFragment_(fragment) {
+    this.ampdoc_.win.location.replace(TEXT_FRAGMENT_PREFIX + fragment);
```

Comment:
This will overwrite any current hash fragment. Can we append if it's already present?

```js
if (location.hash) {
  replace(`${location.hash}:~:${fragment}`);
} else {
  replace(`#:~:${fragment}`);
}